URI Reading Error

I have the following Errormessage when i start my Application. 

`2024-01-31T09:56:18.171 UTC ERROR [main|org.eclipse.milo.opcua.stack.core.util.ManifestUtil:98] uris() failed
java.net.URISyntaxException: Illegal character in opaque part at index 30: jar:nested:/C:/Users/.../..../target/.../!BOOT-INF/classes/!/META-INF/MANIFEST.MF
        at java.base/java.net.URI$Parser.fail(URI.java:2974)
        at java.base/java.net.URI$Parser.checkChars(URI.java:3145)
        at java.base/java.net.URI$Parser.parse(URI.java:3181)
        at java.base/java.net.URI.<init>(URI.java:623)
        at java.base/java.net.URL.toURI(URL.java:1056)
        at org.eclipse.milo.opcua.stack.core.util.ManifestUtil.uris(ManifestUtil.java:93)
        at org.eclipse.milo.opcua.stack.core.util.ManifestUtil.load(ManifestUtil.java:46)
        at org.eclipse.milo.opcua.stack.core.util.ManifestUtil.<clinit>(ManifestUtil.java:33)
        at org.eclipse.milo.opcua.sdk.client.OpcUaClient.<clinit>(OpcUaClient.java:141)
        at ....createClient(CoreConnection.java:239)
        at ....initializeConnection(CoreConnection.java:800)
        at ....doConnect(CoreConnection.java:783)
        at ....openConnection(AbstractCoreConnection.java:42)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:352)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:242)
        at jdk.proxy2/jdk.proxy2.$Proxy98.openConnection(Unknown Source)
        at ....afterPropertiesSet(DelegatingProcessVariableObservable.java:35)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1822)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1771)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:601)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:523)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:325)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:323)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1441)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1348)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.resolveFieldValue(AutowiredAnnotationBeanPostProcessor.java:769)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:752)
        at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:145)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:493)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1420)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:523)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:325)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:323)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1441)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1348)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.resolveFieldValue(AutowiredAnnotationBeanPostProcessor.java:769)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:752)
        at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:145)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:493)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1420)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:523)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:325)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:323)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:973)
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:946)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:616)
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:753)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:455)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:323)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1342)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1331)
        at ....ApplicationStarter.main(ApplicationStarter.java:19)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:91)
        at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:53)
        at org.springframework.boot.loader.launch.PropertiesLauncher.main(PropertiesLauncher.java:574)` Even tho it thwrows this Error the connection to the OPC-UA Server seems to work perfectly fine. I checked evry uri twice but it always throws this Message.  Here is the Code-Part that i use.   
`OpcUaClient opcUaClient = OpcUaClient.create(
                    this.endpointUrl,
                    endpoints -> endpoints.stream().filter(this.endpointFilter()).findFirst(),
                    configBuilder -> buildConfig(configBuilder, loader, certificateValidator));`
this.endpointUrl looks like this: opcua.endpointurl=opc.tcp://ip:port

> You would need to check out this branch locally, build and install a Milo dev version via `mvn clean install`, then update your version to point at `0.6.13-SNAPSHOT`.

The 0.6.13-SNAPSHOT has a invalid jar and pom. When i build the project i have the following error: `[ERROR] Tests run: 195, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 9.256 s <<< FAILURE! - in TestSuite
[ERROR] org.eclipse.milo.opcua.stack.core.util.validation.CertificateValidationUtilTest.testBuildAndValidate_LeafIntermediateSigned_Revoked  Time elapsed: 0.07 s  <<< FAILURE!
java.lang.AssertionError: expected [StatusCode{name=Bad_CertificateRevoked, value=0x801D0000, quality=bad}] but found [StatusCode{name=Bad_CertificateIssuerRevocationUnknown, value=0x801C0000, quality=bad}]
        at org.eclipse.milo.opcua.stack.core.util.validation.CertificateValidationUtilTest.testBuildAndValidate_LeafIntermediateSigned_Revoked(CertificateValidationUtilTest.java:232)`

So is build it with -DskipTests. But the jar and pom of version 0.6.13-SNAPSHOT are somehow invalid to request them in my project. When i try 0.6.12 the error that i mentioned in this issue still occures.

--- a/opc-ua-stack/stack-core/src/main/java/org/eclipse/milo/opcua/stack/core/util/ManifestUtil.java
+++ b/opc-ua-stack/stack-core/src/main/java/org/eclipse/milo/opcua/stack/core/util/ManifestUtil.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019 the Eclipse Milo Authors
+ * Copyright (c) 2024 the Eclipse Milo Authors
  *
  * This program and the accompanying materials are made
  * available under the terms of the Eclipse Public License 2.0
@@ -10,16 +10,12 @@
 
 package org.eclipse.milo.opcua.stack.core.util;
 
-import java.io.IOException;
 import java.io.InputStream;
-import java.net.URI;
 import java.net.URL;
-import java.util.Collections;
 import java.util.Enumeration;
-import java.util.HashSet;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
-import java.util.Set;
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 
@@ -30,38 +26,40 @@ import org.slf4j.LoggerFactory;
 public class ManifestUtil {
 
     private static final Logger logger = LoggerFactory.getLogger(ManifestUtil.class);
-    private static final Map<String, String> attributes = load();
+    private static final Map<String, String> ATTRIBUTES = readAttributes();
 
     public static Optional<String> read(String name) {
-        return Optional.ofNullable(attributes.get(name));
+        return Optional.ofNullable(ATTRIBUTES.get(name));
     }
 
     public static boolean exists(String name) {
-        return attributes.containsKey(name);
+        return ATTRIBUTES.containsKey(name);
     }
 
-    private static Map<String, String> load() {
+    private static Map<String, String> readAttributes() {
         Map<String, String> attributes = Maps.newConcurrentMap();
 
-        for (URI uri : uris()) {
-            try {
-                attributes.putAll(load(uri.toURL()));
-            } catch (Throwable t) {
-                logger.error("load(): '{}' failed", uri, t);
+        try {
+            Enumeration<URL> resources = ManifestUtil.class
+                .getClassLoader()
+                .getResources("META-INF/MANIFEST.MF");
+
+            while (resources.hasMoreElements()) {
+                URL url = resources.nextElement();
+
+                attributes.putAll(readAttributes(url));
             }
+        } catch (Throwable t) {
+            logger.error("readAttributes() failed", t);
         }
 
         return attributes;
     }
 
-    private static Map<String, String> load(URL url) throws IOException {
-        return load(url.openStream());
-    }
+    private static Map<String, String> readAttributes(URL url) {
+        Map<String, String> props = new HashMap<>();
 
-    private static Map<String, String> load(InputStream stream) {
-        Map<String, String> props = Maps.newConcurrentMap();
-
-        try {
+        try (InputStream stream = url.openStream()) {
             Manifest manifest = new Manifest(stream);
             Attributes attributes = manifest.getMainAttributes();
 
@@ -70,34 +68,10 @@ public class ManifestUtil {
                 props.put(key.toString(), value);
             }
         } catch (Throwable t) {
-            logger.error("#load(): failed", t);
-        } finally {
-            try {
-                stream.close();
-            } catch (IOException e) {
-                // ignored
-            }
+            logger.error("readAttributes(): '{}' failed", url, t);
         }
 
         return props;
-    }
-
-    private static Set<URI> uris() {
-        try {
-            Enumeration<URL> resources = ManifestUtil.class
-                .getClassLoader()
-                .getResources("META-INF/MANIFEST.MF");
-
-            Set<URI> uris = new HashSet<>();
-            while (resources.hasMoreElements()) {
-                uris.add(resources.nextElement().toURI());
-            }
-
-            return uris;
-        } catch (Throwable t) {
-            logger.error("uris() failed", t);
-            return Collections.emptySet();
-        }
     }
 
 }

